### ISSUE TITLE
Add serialization support to ufo3 branch.

### ISSUE THREAD

**Issue by typesupply** [typesupply/defcon#23]
There needs to be a simple base implementation of object serialization in the objects. This needs to be easy to extend and subclass.

**Comment by adrientetar** [comment]
I might work on this, I need a serialization system that works for undo as well as clipboard (e.g. for clipboard, I don’t want to copy naming data).

@typesupply I’m wondering, why doesn’t base implementation of getDataToSerialize just return `self.__dict__`?  
(Then subclasses would remove from the dict instead of adding to it.) 
This means that when adding additional data fields to classes, they are picked automatically by the serialization system. 
Do you have any particular idea/requirement for such a system?

**Comment by adrientetar** [comment]
Okay, I guess returning self.__dict__ would include all the object’s functions which might be why it isn’t done so.

**Comment by typesupply** [comment]
Re: self.__dict__: Right. The returned dict may also contain mutable data.

This is a big, gnarly problem to solve and I've deliberately ignored it because it can get app specific very fast and that sucks for everyone but the app. (The serialization stuff in the master branch worked GREAT for my needs but was terrible for everyone else. :) So, this needs to be very easy to customize and it needs to be very lightweight. Some half-baked thoughts...

I think that the serialized objects should be strings so that they take up as little memory as possible. That requires some read/write functionality and that will require some knowledge of everything that needs to be read/written. That's tough because every app is going to have its own unique attributes and objects. I've considered using pickle to pickle entire trees of objects. The problem with that is that every object needs to be pickle-able and that's not a certainty. Still, pickle is probably the easiest and most efficient way to implement this. Maybe we could implement something like this:

```python
# Base

def serialize(self):
    data = self.getDataForSerialization()
    return cPickle.dumps(data)

def deserialize(self, data):
    data = cPickle.loads(data)
    self.setDataFromSerialization(data)

def getDataForSerialization(self):
    """
    Return a dict of data that can be pickled.
    """
    return {}

def setDataFromSerialization(self, data):
    """
    Restore data from the provided dict.
    """
    pass

# Glyph

def getDataForSerialization(self):
    data = dict(
        name=self.name,
        unicodes=self.unicodes,
        width=self.width,
        height=self.height,
        image=None,
        contours=[contour.getDataForSerialization() for contour in self.contours],
        components=[component.getDataForSerialization() for component in self.components],
        # etc.
    )
    image = self.image
    if image is not None:
        data["image"] = self.image.getDataForSerialization()
    # etc.
    return data

def setDataFromSerialization(self, data):
    self.name = data["name"]
    self.unicodes = data["unicodes"]
    self.contours.clear()
    # etc.
    for contourData in data["contours"]:
        contour = self.instantiateContour()
        contour.setDataFromSerialization(contourData)
    # etc.

# MyGlyph

def getDataForSerialization(self):
    data = super(MyGlyph, self).getDataForSerialization()
    data["myCustomThing"] = self.myCustomThing
    return data

def setDataFromSerialization(self, data):
    super(MyGlyph, self).setDataFromSerialization(data)
    self.myCustomThing = data["myCustomThing"]
```

The time needed for sub-object serialization data creation in could be cumbersome. Representations may be able to help with this.

@typemytype may have some thoughts on this.

**Comment by graphicore** [comment]
I implemented something very similar for the ufo3 branch based on the python3 branch of @adrientetar.
If you like it we can backport it.

Here is the code in the base glyph:
 https://github.com/graphicore/defcon/blob/serialization-py3-ufo-3/Lib/defcon/objects/base.py#L326

Here is a comparison of all changes that omits the changes of the python3-ufo3 branch

https://github.com/graphicore/defcon/compare/python3-ufo3...graphicore:serialization-py3-ufo-3

`defcon.objects.info.Info` was on a diet, so I could use the `_properties` class-property to know which values need serialization

https://github.com/graphicore/defcon/blob/serialization-py3-ufo-3/Lib/defcon/objects/info.py

We should eventually support both python3 and python2 in the ufo3 branch, so it's easier to merge stuff.

This is not tested in depth, but I think it is correct and there are no big bugs.  I did not really care to fiddle with the events/observer stuff, I think right now all actions upon deserialization will be reported. Which is kind of the default, but maybe not desirable (?)

The serialize function takes `dumpFunc` as optional argument and deserialize symetrically takes `loadFunc` the defaults are `pickle.dumps` and `pickle.loads` . json.dumps / json.loads works, there is no validation of the data or such, so I wouldn't call it a ufo-json. It's just a dump of defcon state.

serialize also has `whitelist` and `blacklist` to control what is loaded. So, to remove 'unicode' and 'name' from the dump of a glyph:  glyph.serialize(blacklist=['unicode', 'name'])
